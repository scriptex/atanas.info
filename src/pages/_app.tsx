import type { AppProps } from 'next/app';
import { FC, useState, useEffect, createContext } from 'react';

import { Head } from '@components';
import { Theme, onThemeChange, setThemeClassName } from '@scripts/shared';

import '@styles/index.css';

export const AppContext = createContext({
	contactVisible: false,
	setContactVisible: (state: boolean) => state
});

export const App: FC<AppProps> = ({ Component, pageProps }: AppProps) => {
	const [contactVisible, setContactVisible] = useState(false);
	const value: any = { contactVisible, setContactVisible };

	useEffect(() => {
		setThemeClassName(window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light');

		onThemeChange(({ media, matches }) => {
			if (!matches) {
				return;
			}

			setThemeClassName(media.replace(/^\(prefers-color-scheme: (.*)\)$/, (_, match) => match) as Theme);
		});

		if (CSS && CSS.paintWorklet && CSS.paintWorklet.addModule && typeof CSS.paintWorklet.addModule === 'function') {
			CSS.paintWorklet.addModule(new URL('/houdini/avatar-polygon.js', window.location.href));
			CSS.paintWorklet.addModule(new URL('/houdini/blotto.js', window.location.href));
			CSS.paintWorklet.addModule(new URL('/houdini/bubbles.js', window.location.href));
			CSS.paintWorklet.addModule(new URL('/houdini/bytemare.js', window.location.href));
			CSS.paintWorklet.addModule(new URL('/houdini/circles.js', window.location.href));
			CSS.paintWorklet.addModule(new URL('/houdini/confetti.js', window.location.href));
			CSS.paintWorklet.addModule(new URL('/houdini/parallelowow.js', window.location.href));
			CSS.paintWorklet.addModule(new URL('/houdini/slanted-backgrounds.js', window.location.href));
		}

		import('react-gtm-module').then(({ default: TagManager }) => {
			TagManager.initialize({
				gtmId: process.env.NEXT_PUBLIC_GTM_ID || ''
			});
		});
	}, []);

	return (
		<>
			<Head />

			<AppContext.Provider value={value}>
				<Component {...pageProps} />
			</AppContext.Provider>
		</>
	);
};

export default App;
