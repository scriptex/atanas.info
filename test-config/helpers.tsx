import { act } from 'react-dom/test-utils';
import type { FC } from 'react';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '@scripts/shared';

type ComponentType = FC<Readonly<React.ComponentProps<any>>>;

export const test = async (Component: ComponentType): Promise<RenderResult> => {
	let result!: RenderResult;

	await act(async () => {
		result = await waitFor(() => render(<Component />));
	});

	return result;
};

export const snapshotTest = (Component: ComponentType, buttonSelector?: string): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			const { asFragment, container } = await test(Component);

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				await act(async () => {
					await container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			const { unmount, asFragment } = await test(Component);

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};

export const mockAudioContext = (): jest.Mock => {
	const mockConnect = jest.fn();

	const mockgetByteFrequencyData = jest.fn();

	const mockcreateMediaElementSource = jest.fn(() => ({
		connect: mockConnect
	}));

	const mockcreateAnalyser = jest.fn(() => ({
		connect: mockConnect,
		frequencyBinCount: [0, 1, 2],
		getByteFrequencyData: mockgetByteFrequencyData
	}));

	const mockcreateOscillator = jest.fn(() => ({
		channelCount: 2
	}));

	const mockChannelSplitterConnect = jest.fn(n => n);

	const mockcreateChannelSplitter = jest.fn(() => ({
		connect: mockChannelSplitterConnect
	}));

	const mockedAudioContext = jest.fn(() => ({
		createAnalyser: mockcreateAnalyser,
		createOscillator: mockcreateOscillator,
		createChannelSplitter: mockcreateChannelSplitter,
		createMediaElementSource: mockcreateMediaElementSource
	}));

	return mockedAudioContext;
};

export const mockFetch = (value: Record<string, string | boolean>): void => {
	global.fetch = jest.fn().mockImplementation(() =>
		Promise.resolve({
			json: () => Promise.resolve(value)
		})
	);
};
