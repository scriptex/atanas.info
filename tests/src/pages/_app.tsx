import type { AppProps } from 'next/app';

import { App } from '@pages/_app';
import { snapshotTest } from '@test-config/helpers';

const env = process.env;
const props = {
	Component: () => <div />,
	pageProps: {}
} as unknown as AppProps;

beforeEach(() => {
	jest.resetModules();
	process.env = { ...env, NEXT_PUBLIC_GTM_ID: 'test-id' };
});

afterEach(() => {
	process.env = env;
});

window.CSS = window.CSS || {
	paintWorklet: {
		addModule: path => path
	}
};

snapshotTest(() => <App {...props} />);
