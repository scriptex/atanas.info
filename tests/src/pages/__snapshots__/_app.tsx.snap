// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` component Should render the  component 1`] = `
<DocumentFragment>
  <div />
  Vercel Analytics
</DocumentFragment>
`;

exports[` component Should unmount the  component 1`] = `<DocumentFragment />`;
